libs/@local/hash-graph-sdk/typescript/src/entity.ts

```diff
@@ -103,6 +104,44 @@ const isGraphApiEntity = <Properties extends EntityPropertiesObject>(
   );
 };
 
+export const flattenedPropertyMetadataMap = (
+  metadata: PropertyMetadataObject,
+): {
+  path: PropertyPath;
+  metadata: Required<PropertyMetadataElement>["metadata"];
+}[] => {
+  const flattened: {
+    path: PropertyPath;
+    metadata: Required<PropertyMetadataElement>["metadata"];
+  }[] = [];
+
+  const visitElement = (
+    path: PropertyPath,
+    element: PropertyMetadataElement,
+  ): void => {
+    if ("properties" in element) {
+      for (const [key, value] of typedEntries(element.properties)) {
+        visitElement([...path, key], value);
+      }
+    } else if ("elements" in element) {
+      for (const [index, value] of element.elements.entries()) {
+        visitElement([...path, index], value);
+      }
+    }
```

Comment:
The general handling is easier if `PropertyMetadataElement` has different properties. This PR uses `properties` for objects, `elements` for arrays, and `value` for leafs. Another way would be always using `value` but using `type: "..."`. It's also possible to completely omit `type` but with a runtime cost as `isArray` would need to be used.

Reply:
I tried the `value` approach without `type` and it's not as bad as expected: [`0e3533f` (#4563)](https://github.com/hashintel/hash/pull/4563/commits/0e3533f422f2e93c20c44fa8f9cd0535af0e556c)